### PR TITLE
Variable methods to free functions

### LINKED ISSUE #1099: Make reshape, transpose, ... free functions
Some shape-related operations are defined as members of `Variable`. Change to free functions for consistency.

### PR DESCRIPTION
**Pull request by OwenArnold**
Fixes #1099 and #1255

* `reshape`, `transpose` made into free functions. Due to some encapsulation issues. `transpose` remains a public member on `VariableConstView` and `VariableView`
* `transpose` exposed to python
* new `shape.cpp` for `transpose`, `reshape` and `concatenate`

### CODE REVIEW

File: python/variable.cpp [hunk 1]
```diff
@@ -290,7 +290,7 @@ Mostly equivalent to Variable, see there for details.)");
       [](const VariableView &self, const std::vector<Dim> &labels,
```
Comment (SimonHeybrock) [inline python/variable.cpp:290]: Should probably also having bindings for `Variable`, to avoid copies?

File: variable/shape.cpp [hunk 2]
```diff
@@ -138,4 +138,25 @@ Variable reshape(const VariableConstView &view, const Dimensions &dims) {
   return reshaped;
 }
 
+VariableView transpose(Variable &var, const std::vector<Dim> &dims) {
+  return transpose(VariableView(var), dims);
+}
+
+VariableConstView transpose(const Variable &var, const std::vector<Dim> &dims) {
+  return transpose(VariableConstView(var), dims);
+}
```
Comment (SimonHeybrock) [inline variable/shape.cpp:147]: Is this overload needed? Should be able to make use of implicit conversion?

File: python/operations.cpp [hunk 3]
```diff
@@ -119,6 +119,27 @@ template <typename T> void bind_contains_events(py::module &m) {
           .c_str());
 }
 
+template <class T> void bind_reshape(pybind11::module &mod) {
```
Comment (SimonHeybrock) [inline python/operations.cpp:122]: I suppose these (and concatenate) should go into `python/shape.cpp` to match the naming in the other modules. Could you move them? Cheers!

Comment (OwenArnold) [reply python/operations.cpp:122]: Yes. I'm also exposing `transpose` as that's missing.

**Pull request merged**